# Phase 2 — Data Loading

## Objective
The objective of this phase is to load the raw loan transactions and economic indicators datasets and perform a comprehensive inspection. This includes checking dataset shapes, previewing records, identifying data types, assessing memory usage, computing descriptive statistics, summarizing missing values, identifying duplicates, and analyzing the distribution of the target variable.

### Import Libraries and Load Data
We begin by importing the necessary packages and loading our datasets: `train.csv` (contains customer loan details) and `economic_indicators.csv` (contains macroeconomic features like inflation and exchange rates).

In [ ]:
# Import necessary libraries for data manipulation and system operations
import os
import pandas as pd
import sys
sys.path.append('..')
# Import custom utility function to load raw datasets
from src.utils import load_raw_data

# Define path to raw data directory
raw_dir = r'../data/raw'
# Load both training dataset (loan transactions) and economic indicators dataset
train_df, econ_df = load_raw_data(raw_dir)


### Dataset Shape and Preview
Let us inspect the dimensions of the datasets and preview a few rows to understand their structure.

In [ ]:
# Display the shape (dimensions) of both datasets
print(f'Train Data Shape: {train_df.shape}')
print(f'Economic Indicators Shape: {econ_df.shape}')
# Preview the first 3 rows of each dataset to understand structure and content
display(train_df.head(3))
display(econ_df.head(3))


Train Data Shape: (68654, 16)
Economic Indicators Shape: (27, 25)


,ID,customer_id,country_id,tbl_loan_id,lender_id,loan_type,Total_Amount,Total_Amount_to_Repay,disbursement_date,due_date,duration,New_versus_Repeat,Amount_Funded_By_Lender,Lender_portion_Funded,Lender_portion_to_be_repaid,target
0,ID_257244237466267278,257244,Kenya,237466,267278,Type_1,16477.0,16642.0,2022-08-15,2022-08-22,7,Repeat Loan,2436.0,0.147842,2460.0,0
1,ID_249619214444267278,249619,Kenya,214444,267278,Type_1,5599.0,5599.0,2022-07-13,2022-07-20,7,Repeat Loan,2799.5,0.500000,2800.0,0
2,ID_251457257838267278,251457,Kenya,257838,267278,Type_1,4679.0,4746.0,2022-09-12,2022-09-19,7,Repeat Loan,7.8,0.001667,8.0,0


,Country,Indicator,YR2001,YR2002,YR2003,YR2004,YR2005,YR2006,YR2007,YR2008,...,YR2014,YR2015,YR2016,YR2017,YR2018,YR2019,YR2020,YR2021,YR2022,YR2023
0,Ghana,"Inflation, consumer prices (annual %)",41.509496,9.360932,29.772980,18.042739,15.438992,11.679184,10.734267,16.494640,...,15.489616,17.149970,17.454635,12.371922,7.808765,7.143640,9.887290,9.971089,31.255895,38.106966
1,Cote d'Ivoire,"Inflation, consumer prices (annual %)",4.361529,3.077265,3.296807,1.457988,3.885830,2.467191,1.892006,6.308528,...,0.448682,1.251500,0.723178,0.685881,0.359409,-1.106863,2.425007,4.091952,5.276167,4.387117
2,Kenya,"Inflation, consumer prices (annual %)",5.738598,1.961308,9.815691,11.624036,10.312778,14.453734,9.758880,26.239817,...,6.878155,6.582154,6.297250,8.005650,4.689806,5.239638,5.405162,6.107936,7.659863,7.671396


### Data Types and Memory Usage
Checking data types and memory usage is critical for identifying columns that require conversion (e.g., date columns) and managing performance.

In [ ]:
# Display detailed information about training dataset (data types and memory usage)
print('--- Train Data Info ---')
train_df.info()
# Display detailed information about economic indicators dataset
print('\n--- Economic Indicators Info ---')
econ_df.info()


--- Train Data Info ---
<class 'pandas.DataFrame'>
RangeIndex: 68654 entries, 0 to 68653
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID                           68654 non-null  str    
 1   customer_id                  68654 non-null  int64  
 2   country_id                   68654 non-null  str    
 3   tbl_loan_id                  68654 non-null  int64  
 4   lender_id                    68654 non-null  int64  
 5   loan_type                    68654 non-null  str    
 6   Total_Amount                 68654 non-null  float64
 7   Total_Amount_to_Repay        68654 non-null  float64
 8   disbursement_date            68654 non-null  str    
 9   due_date                     68654 non-null  str    
 10  duration                     68654 non-null  int64  
 11  New_versus_Repeat            68654 non-null  str    
 12  Amount_Funded_By_Lender      68654 non-null  float64
 13  Len

### Descriptive Statistics
We run descriptive statistics to examine the central tendency, dispersion, and range of our numerical features.

In [ ]:
# Compute and display descriptive statistics for numerical columns in training dataset
print('--- Train Data Numerical Stats ---')
display(train_df.describe())
# Compute and display descriptive statistics for numerical columns in economic indicators
print('--- Economic Indicators Numerical Stats ---')
display(econ_df.describe())


--- Train Data Numerical Stats ---


,customer_id,tbl_loan_id,lender_id,Total_Amount,Total_Amount_to_Repay,duration,Amount_Funded_By_Lender,Lender_portion_Funded,Lender_portion_to_be_repaid,target
count,68654.000000,68654.000000,68654.000000,6.865400e+04,6.865400e+04,68654.000000,6.865400e+04,68654.000000,6.865400e+04,68654.000000
mean,254392.749264,262982.580593,266432.652591,1.459270e+04,1.539934e+04,8.503569,2.501309e+03,0.219135,2.603978e+03,0.018324
std,26623.526913,39328.023195,3567.136062,1.336684e+05,1.611572e+05,13.426630,1.148607e+04,0.129656,1.291656e+04,0.134120
min,145.000000,101323.000000,245684.000000,5.000000e+00,0.000000e+00,1.000000,0.000000e+00,0.000000,0.000000e+00,0.000000
25%,248963.000000,233890.500000,267278.000000,2.289000e+03,2.321000e+03,7.000000,2.390000e+02,0.121316,2.420000e+02,0.000000
50%,255410.000000,260246.500000,267278.000000,5.248000e+03,5.299000e+03,7.000000,9.090000e+02,0.300000,9.270000e+02,0.000000
75%,262332.000000,286898.750000,267278.000000,1.138800e+04,1.159600e+04,7.000000,2.258400e+03,0.300000,2.303000e+03,0.000000
max,312737.000000,375320.000000,267278.000000,2.000000e+07,2.415284e+07,1096.000000,1.600000e+06,1.168119,1.821338e+06,1.000000


--- Economic Indicators Numerical Stats ---


,YR2001,YR2002,YR2003,YR2004,YR2005,YR2006,YR2007,YR2008,YR2009,YR2010,...,YR2014,YR2015,YR2016,YR2017,YR2018,YR2019,YR2020,YR2021,YR2022,YR2023
count,20.000000,20.000000,20.000000,20.000000,24.000000,24.000000,24.000000,24.000000,24.000000,24.000000,...,24.000000,21.000000,21.000000,21.000000,17.000000,17.000000,17.000000,17.000000,14.000000,14.000000
mean,210.842603,206.489954,201.026417,196.937123,164.782692,163.548217,161.936892,161.543137,162.338789,163.808707,...,164.375498,188.461056,189.086003,188.011613,229.192362,230.834145,230.824774,229.991393,60.578945,61.707586
std,415.661401,414.344484,407.896288,406.192474,376.384006,376.638599,374.801141,373.276370,374.282893,375.003697,...,374.796407,402.329144,402.171491,401.833615,436.561975,438.135892,437.353611,436.170093,164.827016,160.847507
min,0.716305,0.792417,0.866764,0.899495,-0.558333,-8.009867,-3.061667,-3.140000,-10.096004,-3.273333,...,-1.920000,-1.844167,-1.630000,-1.350000,0.359409,-1.106863,2.425007,2.590000,2.434000,2.403000
25%,6.414025,5.508667,7.256000,5.708814,5.803748,4.385458,4.802235,5.102542,4.664309,4.408372,...,4.127750,3.714642,5.237000,4.350533,4.585325,5.014000,5.405162,5.693000,5.408375,4.736109
50%,18.739167,15.942642,13.661172,13.078333,8.981250,8.691124,8.536963,9.999904,9.035825,10.273762,...,8.311229,6.897669,7.871005,6.490000,8.292700,7.506070,6.968703,7.427581,7.402519,8.419543
75%,50.772921,48.073822,46.555495,47.483551,28.444069,26.845454,29.629276,30.417882,28.538951,29.843078,...,33.005013,17.149970,17.454635,14.062500,101.301574,101.991298,106.450780,109.637747,12.264902,13.045543
max,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,...,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,1348.000000,623.759701,606.569750


### Missing Values and Duplicates Summary
We inspect our datasets for missing values and duplicate rows, which will dictate our cleaning strategy.

In [ ]:
# Check for missing values in training dataset
print('Train Data Missing Values:')
print(train_df.isnull().sum())
# Count the number of duplicate rows in training dataset
print('\nTrain Data Duplicates:', train_df.duplicated().sum())
# Check for missing values in economic indicators dataset
print('\nEconomic Indicators Missing Values:')
print(econ_df.isnull().sum())


Train Data Missing Values:
ID                             0
customer_id                    0
country_id                     0
tbl_loan_id                    0
lender_id                      0
loan_type                      0
Total_Amount                   0
Total_Amount_to_Repay          0
disbursement_date              0
due_date                       0
duration                       0
New_versus_Repeat              0
Amount_Funded_By_Lender        0
Lender_portion_Funded          0
Lender_portion_to_be_repaid    0
target                         0
dtype: int64



Train Data Duplicates: 0

Economic Indicators Missing Values:
Country       0
Indicator     0
YR2001        7
YR2002        7
YR2003        7
YR2004        7
YR2005        3
YR2006        3
YR2007        3
YR2008        3
YR2009        3
YR2010        3
YR2011        3
YR2012        3
YR2013        3
YR2014        3
YR2015        6
YR2016        6
YR2017        6
YR2018       10
YR2019       10
YR2020       10
YR2021       10
YR2022       13
YR2023       13
dtype: int64


### Unique Value Counts
Let's check the unique value counts for key columns to understand cardinality.

In [ ]:
# Iterate over key categorical columns and display the number of unique values in each
for col in ['country_id', 'loan_type', 'New_versus_Repeat', 'lender_id']:
    print(f'Unique values in {col}: {train_df[col].nunique()}')


Unique values in country_id: 1
Unique values in loan_type: 19
Unique values in New_versus_Repeat: 2
Unique values in lender_id: 4


### Target Class Distribution
Understanding target class distribution is vital, especially for credit default prediction where defaults (1) are typically rare compared to repayments (0).

In [ ]:
# Count the frequency of each class (0=paid, 1=default) in the target variable
target_counts = train_df['target'].value_counts()
# Calculate the percentage distribution of target classes
target_pct = train_df['target'].value_counts(normalize=True) * 100
# Display absolute counts of each target class
print('Target Counts:')
print(target_counts)
# Display percentage distribution of target classes
print('\nTarget Percentages:')
print(target_pct)


Target Counts:
target
0    67396
1     1258
Name: count, dtype: int64

Target Percentages:
target
0    98.167623
1     1.832377
Name: proportion, dtype: float64


### Summary of Findings
- **Training set size**: 68,654 records, 16 features.
- **Economic indicators**: 27 rows (3 countries: Ghana, Cote d'Ivoire, Kenya; 9 indicators) and years ranging from 2001 to 2023.
- **Missing values**: The training set has no missing values. The economic indicators dataset has several missing values for Kenya (specifically average precipitation and fossil fuel consumption).
- **Target imbalance**: Highly imbalanced dataset: ~98.17% paid (0) vs ~1.83% default (1). This dictates that we must use metrics like F1-score, Recall, and Precision, and use class balancing techniques during training.